### PR TITLE
[FX] Fix tracing support for ScriptMethods

### PR DESCRIPTION
**Pull request by jamesr66a**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52884 [FX] Fix tracing support for ScriptMethods**

Previously, TorchBind method invocations actually weren't captured by symbolic tracing because TorchBind methods were implemented as `ScriptMethod`s. `ScriptMethods` did NOT participate in `__torch_function__` dispatch, thus those calls weren't actually recorded. The original `test_torchbind_class_attribute_in_fx` was a red herring, since the method called did not take any tensor-valued arguments (it took no arguments at all) and the test checked for correctness based on value. The value captured was actually just a baked-in constant.

This fixes all of that by adding `__torch_function__` dispatch support for `ScriptMethod`s. It also adds a test that calls a method that takes a `Tensor` to make sure this actually works

Differential Revision: [D26675801](https://our.internmc.facebook.com/intern/diff/D26675801)

### CODE REVIEW

**Comment by jamesr66a** [comment]
@hameerabbasi @ezyang Can you take a look at this? I'm making it so that TorchScript callables (in this patch, method calls) are admissible to `__torch_function__` dispatch. This allows us to record TorchBind method invocations in FX, for one. I'm particularly interested in if there are deeper implications here wrt the design of `__torch_function__`. Please let me know!

File: torch/csrc/jit/python/pybind_utils.h
Comment (hameerabbasi) [inline torch/csrc/jit/python/pybind_utils.h:968]: This will fail for `TensorList` operands.

File: torch/csrc/jit/python/pybind_utils.h
Comment (hameerabbasi) [inline torch/csrc/jit/python/pybind_utils.h:982]: Same here, `TensorList` will fail. You may need to add an explicit check for `isinstance(arg, (tuple, list))` and recurse inside too.

File: torch/csrc/jit/python/pybind_utils.h
Comment (hameerabbasi) [inline torch/csrc/jit/python/pybind_utils.h:967]: Perhaps switch to a range-based for over `args_vec`? It's both faster and more readable.

File: torch/csrc/jit/python/pybind_utils.h
Comment (hameerabbasi) [inline torch/csrc/jit/python/pybind_utils.h:995]: Types need to be weakly ordered -- Subclasses before superclasses. That invariant isn't satisfied here, unfortunately.

Comment (hameerabbasi) [reply torch/csrc/jit/python/pybind_utils.h:995]: Scratch that -- Looks like you aren't using `overloaded_types` at all.

**Review by hameerabbasi — CHANGES_REQUESTED** [review]
This will require updates to the RFC -- Currently the only methods we support are on `Tensor`. We might need a small update to say `ScriptMethod`s are passed in as `func` since they're callable.

We also might need a testing script to test `ScriptMethod`s.

File: torch/csrc/jit/python/pybind_utils.h
Comment (jamesr66a) [inline torch/csrc/jit/python/pybind_utils.h:1006]: @hameerabbasi To be clear, this function is a direct copy-paste from the implementation for custom ops: https://github.com/pytorch/pytorch/blob/1d6bd157902d4b1347a5d03122d02b407658e263/torch/csrc/jit/python/init.cpp#L1020

So any deficiencies you're pointing out also apply there and should probably be split off into issues, as this points out deficiencies in the current implementation of `__torch_function__`.

**Comment by jamesr66a** [comment]
@hameerabbasi please let me know how to move forward on this. Is updating the RFC/making the final call an action item on your side?

File: torch/csrc/jit/python/pybind_utils.h
Comment (hameerabbasi) [reply torch/csrc/jit/python/pybind_utils.h:1006]: Code replication is usually bad -- We should think about extracting this somewhere so there aren't 4+ implementations to change when we fix these issues:

- Main C++ one
- This one
- The one you refer to in your comment
- The Python one
- Did I miss any? 🙂 

In any case, we can split it out into issues; but we should deduplicate first.

File: torch/csrc/jit/python/script_init.cpp
Comment (ezyang) [inline torch/csrc/jit/python/script_init.cpp:1102]: Is there a way to easily raise an error if you hit the TODO case? Maybe not...

**Comment by hameerabbasi** [comment]
I may have misunderstood -- Does "method" here mean in the Python sense (instance method) or is it a regular callable?

**Comment by ezyang** [comment]
It's an honest to goodness method. What's going on here is kind of weird but basically it's like if you had:

```
class RELu:
  def run(t: Tensor):
    if has_torch_function_unary(t):
      return handle_torch_function(run, self, t)
    return t.relu()
```

I actually think this is pretty analogous to Tensor methods, because we don't really use bound methods in the API, I think.

File: torch/csrc/jit/python/pybind_utils.h
Comment (jamesr66a) [reply torch/csrc/jit/python/pybind_utils.h:1006]: @hameerabbasi If it's okay, I'd prefer to land something like this PR before deduplicating. Lack of support for these methods is blocking some important use cases now. IMO landing duplicated code now should be OK to unblock so long as we make sure that someone goes in to deduplicate afterward, and make sure nobody's diverging the implementations in the intervening time. Does that sound good to you?

File: torch/csrc/jit/python/pybind_utils.h
Comment (hameerabbasi) [reply torch/csrc/jit/python/pybind_utils.h:995]: `overloaded_types` can be removed, I believe.

File: torch/csrc/jit/python/pybind_utils.h
Comment (hameerabbasi) [reply torch/csrc/jit/python/pybind_utils.h:968]: Can this take `TensorList`s? I believe the reason to skip in the other place was simply that they didn’t need it.

File: torch/csrc/jit/python/pybind_utils.h
Comment (hameerabbasi) [reply torch/csrc/jit/python/pybind_utils.h:1006]: It sounds good, I’ll open an issue for that tomorrow.

**Comment by hameerabbasi** [comment]
Then we need alterations to the RFC and test infrastructure. I believe there are some places  in the tests that assume `self` is a `Tensor`.

File: torch/csrc/jit/python/pybind_utils.h
Comment (esqu1) [reply torch/csrc/jit/python/pybind_utils.h:1006]: @hameerabbasi What James said above; we actually really need this functionality internally for enabling some of our models to be FX+TorchScript compatible. It would be great if we could have this landed this soon and then revisit the refactoring in a later PR.

Comment (hameerabbasi) [reply torch/csrc/jit/python/pybind_utils.h:1006]: I won't block this PR based on that, I'll comment here once the issue is open.

File: torch/csrc/jit/python/pybind_utils.h
Comment (hameerabbasi) [reply torch/csrc/jit/python/pybind_utils.h:968]: I see from [this comment](https://github.com/pytorch/pytorch/pull/52884#issuecomment-788128434) that no `TensorList`s are needed, if so, you can resolve this, @jamesr66a.

File: torch/csrc/jit/python/pybind_utils.h
Comment (hameerabbasi) [reply torch/csrc/jit/python/pybind_utils.h:982]: As well as this one.

**Comment by hameerabbasi** [comment]
I'd only like small changes here, considering [this comment](https://github.com/pytorch/pytorch/pull/52884#discussion_r585487701) is true. Specifically, moving to iterating over the vector and removing the types.

**Comment by hameerabbasi** [comment]
For this to get my approval, I'd like:

- Iteration over `args_vec` instead of `args`: https://github.com/pytorch/pytorch/pull/52884#discussion_r584152093
- Removal of `overloaded_type` computation: https://github.com/pytorch/pytorch/pull/52884#discussion_r584975897
- Verification that the `TensorList` case will never be hit: https://github.com/pytorch/pytorch/pull/52884#discussion_r584151712 and https://github.com/pytorch/pytorch/pull/52884#discussion_r584151855

Aside from this, it's all good to go on my side.

File: torch/csrc/jit/python/pybind_utils.h
Comment (jamesr66a) [reply torch/csrc/jit/python/pybind_utils.h:967]: I don't know about faster (I would imagine the compiler would be able to do a trivial strength reduction here) but for style I'll change it

File: torch/csrc/jit/python/pybind_utils.h
Comment (jamesr66a) [reply torch/csrc/jit/python/pybind_utils.h:968]: @hameerabbasi  I don't really understand this comment thread.

1) Can you explain exactly what would fail about `TensorList`? I didn't write this code, but from what I can tell, `is_tensor_and_append_overloaded` and `is_tensor_list_and_append_overloaded` are both best-effort and both if `args` is a `Tensor` or if `arg` is a `TensorList` should be covered by the two cases here. Can you explain a bit more about what you're referring to?
2) I don't understand the implication of Ed's comment that suggests that `TensorList` operands are not possible.

Comment (hameerabbasi) [reply torch/csrc/jit/python/pybind_utils.h]: `is_tensor_and_append_overloaded` never checks for `TensorList`, only for something that "looks like a `Tensor`", i.e. has `__torch_function__` and isn't `nn.Parameter` or `Tensor`. See this comment: https://github.com/pytorch/pytorch/blob/a0a1bb074b954338d7f19dac9a9c8136376bce7c/torch/csrc/utils/python_arg_parser.h#L748-L757

If `arg` turns out to be a `TensorList`, then it won't recurse inside the `TensorList` and therefore miss types that have `__torch_function__`.

What I'm looking for here is confirmation that `arg` will never be a `TensorList`, i.e. that the callable that you're overriding here will never take a `TensorList` as a potential argument. Ed's comment has an example of what the callable you're overriding might look like, and if it only takes a `Tensor` in, then we're okay.

Comment (jamesr66a) [reply torch/csrc/jit/python/pybind_utils.h]: @hameerabbasi still don't understand your comment. In this code, there's a call to `is_tensor_list_and_append_overloaded` directly after this comment thread, which AFAICT does do this recursion.

Comment (hameerabbasi) [reply torch/csrc/jit/python/pybind_utils.h]: Ah, missed that one. I apologize for the confusion; consider this resolved.

**Review by hameerabbasi — APPROVED** [review]

File: torch/csrc/jit/python/script_init.cpp
Comment (jamesr66a) [reply torch/csrc/jit/python/script_init.cpp:1102]: I'm actually not sure if UDF `__eq__` is even a thing on `Module`. I could understand it for `Object` but not sure if it makes sense for `Module`

Speaking of which @gmagogsfm can we get a review on this from the TorchScript side? Thanks!

File: test/test_fx.py
Comment (SplitInfinity) [inline test/test_fx.py:1626]: This will probably time out in fbcode (see: D26838029). I suggest using `@unittest.skipIf` instead.

Comment (SplitInfinity) [reply test/test_fx.py:1626]: Actually I think the problem I'm referring to only happens in `test_jit.py`, you can probably ignore this.

**Review by tugsbayasgalan — APPROVED** [review]

**Pull request closed without merging**
